tcpclient.go

```diff
@@ -149,50 +134,98 @@ func (mb *tcpTransporter) Send(aduRequest []byte) (aduResponse []byte, err error
 	mb.mu.Lock()
 	defer mb.mu.Unlock()
 
+	var data [tcpMaxLength]byte
+	recoveryDeadline := time.Now().Add(mb.IdleTimeout)
+
 	// Establish a new connection if not connected
 	if err = mb.connect(); err != nil {
 		return
 	}
-	// Set timer to close when idle
-	mb.lastActivity = time.Now()
-	mb.startCloseTimer()
-	// Set write and read timeout
-	var timeout time.Time
-	if mb.Timeout > 0 {
-		timeout = mb.lastActivity.Add(mb.Timeout)
-	}
-	if err = mb.conn.SetDeadline(timeout); err != nil {
-		return
-	}
-	// Send data
-	mb.logf("modbus: sending % x", aduRequest)
-	if _, err = mb.conn.Write(aduRequest); err != nil {
-		return
-	}
-	// Read header first
-	var data [tcpMaxLength]byte
-	if _, err = io.ReadFull(mb.conn, data[:tcpHeaderSize]); err != nil {
-		return
+	for {
+		// Set timer to close when idle
+		mb.lastActivity = time.Now()
+		mb.startCloseTimer()
+		// Set write and read timeout
+		var timeout time.Time
+		if mb.Timeout > 0 {
+			timeout = mb.lastActivity.Add(mb.Timeout)
+		}
+		if err = mb.conn.SetDeadline(timeout); err != nil {
+			return
+		}
+		// Send data
+		mb.logf("modbus: sending % x", aduRequest)
+		if _, err = mb.conn.Write(aduRequest); err != nil {
+			return
+		}
+		// Read header first
+		if _, err = io.ReadFull(mb.conn, data[:tcpHeaderSize]); err == nil {
+			aduResponse, err = mb.processResponse(data[:])
+			if err == nil && mb.ProtocolRecoveryTimeout > 0 && time.Until(recoveryDeadline) > 0 &&
+				verify(aduRequest, aduResponse) != nil {
+				continue
+			}
+			mb.logf("modbus: received % x\n", aduResponse)
+			return
+			// Read attempt failed
+		} else if (err != io.EOF && err != io.ErrUnexpectedEOF) ||
+			mb.LinkRecoveryTimeout == 0 || time.Until(recoveryDeadline) < 0 {
+			return
+		}
+		mb.logf("modbus: close connection and retry, because of %v", err)
+
+		mb.close()
+		time.Sleep(mb.LinkRecoveryTimeout)
+
+		// Establish a new connection if not connected
+		if err = mb.connect(); err != nil {
+			return
+		}
 	}
+}
+
+func (mb *tcpTransporter) processResponse(aduRequest []byte) (aduResponse []byte, err error) {
```

Comment:
Would you please keep the variable name `data` instead of `aduRequest`? I think `io.ReadFull(mb.conn, aduRequest[:])` is a bit confusing.

Reply:
Good point, I add the feedback in latest commit + it should be buildable with older versions of go now as well. Thanks for the input.